### PR TITLE
The AdaptiveBox

### PR DESCRIPTION
**Pull request by shaobinspire**

### CODE REVIEW

File: Applications/Spire/Source/Ui/AdaptiveBox.cpp [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+  struct SharedWidget {
+    const QWidget* m_widget;
+    PlaceholderLayoutItem* m_controlling_placeholder;
+    std::unordered_set<PlaceholderLayoutItem*> m_items;
+
+    SharedWidget(const QWidget& widget)
+      : m_widget(&widget),
+        m_controlling_placeholder(nullptr) {}
+
+    void register_placeholder(PlaceholderLayoutItem* placeholder) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/AdaptiveBox.cpp:23]: Can the ```placeholder``` be ```nullptr```?

File: Applications/Spire/Source/Ui/AdaptiveBox.cpp [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+      m_items.insert(placeholder);
+    }
+
+    void release_placeholder(PlaceholderLayoutItem* placeholder) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/AdaptiveBox.cpp:27]: See above.

File: Applications/Spire/Source/Ui/AdaptiveBox.cpp [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+      auto item = layout.itemAt(i);
+      if(auto child_layout = item->layout()) {
+        invalidate_layout(*child_layout);
+      }
+    }
+  }
+
+  auto get_min_abs_value_index(const std::vector<std::tuple<bool, int>>& v) {
+    auto candidates = v | std::views::filter([] (auto& value) {
+      return !get<0>(value);
```
Comment (kranar) [inline Applications/Spire/Source/Ui/AdaptiveBox.cpp:293]: Indentation.

File: Applications/Spire/Source/Ui/AdaptiveBox.cpp [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+      replace<QGridLayout, GridLayoutItem>(*grid_layout, parent,
+        [] (QGridLayout& layout) {
+          auto item = GridLayoutItem();
+          layout.getItemPosition(0, &item.m_row, &item.m_column,
+            &item.m_row_span, &item.m_column_span);
+          item.m_item = layout.takeAt(0);
+          return item;
+        },
+        [&] (QGridLayout& layout, GridLayoutItem& item, QWidget& widget) {
+          layout.addItem(make_placeholder(widget, parent, *item.m_item),
```
Comment (kranar) [inline Applications/Spire/Source/Ui/AdaptiveBox.cpp:225]: What is being captured?

File: Applications/Spire/Source/Ui/AdaptiveBox.cpp [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+      });
+    if(std::ranges::empty(candidates)) {
+      return std::distance(v.begin(), std::min_element(v.begin(), v.end(),
+        [] (auto& value1, auto& value2) {
+          return std::abs(get<1>(value1)) < std::abs(get<1>(value2));
+        }));
+    }
+    return std::distance(v.begin(), std::find(v.begin(), v.end(),
+      *std::ranges::min_element(candidates.begin(), candidates.end(),
+      [] (auto& value1, auto& value2) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/AdaptiveBox.cpp:303]: Indent again on line 303.

**Review by kranar — CHANGES_REQUESTED** [review]
This is fantastic.

File: Applications/Spire/Source/Ui/AdaptiveBox.cpp (hunk 4 above)
Comment (shaobinspire) [reply Applications/Spire/Source/Ui/AdaptiveBox.cpp:225]: `parent` is being captured by the lambda expression.

File: Applications/Spire/Source/Ui/AdaptiveBox.cpp [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+    }
+
+    void release_placeholder(PlaceholderLayoutItem& placeholder) {
+      m_items.erase(&placeholder);
+      if(m_items.empty()) {
+        delete m_widget;
+      }
+    }
+
+    bool take_control(PlaceholderLayoutItem* placeholder) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/AdaptiveBox.cpp:34]: Same as above, can the ```placeholder``` be ```nullptr```?

**Pull request merged**